core/src/impl/Kokkos_ViewMapping.hpp

```diff
@@ -3537,110 +3537,90 @@ class ViewMapping<
-    char msg[LEN];
-    char const* const first_part = "View bounds error of view ";
-    char* p                      = msg;
-    char* const end              = msg + LEN - 1;
-    for (char const* p2 = first_part; (*p2 != '\0') && (p < end); ++p, ++p2) {
-      *p = *p2;
-    }
-    for (char const* p2 = label; (*p2 != '\0') && (p < end); ++p, ++p2) {
-      *p = *p2;
-    }
-    *p = '\0';
-    Kokkos::abort(msg);
-  }
-};
-
-/* Check #2: does the ViewMapping have the printable_label_typedef defined?
-   See above that only the non-specialized standard-layout ViewMapping has
-   this defined by default.
-   The existence of this alias indicates the existence of MapType::is_managed
- */
 template <class T>
 using printable_label_typedef_t = typename T::printable_label_typedef;
 
-template <class Map>
-KOKKOS_FUNCTION
-    std::enable_if_t<!is_detected<printable_label_typedef_t, Map>::value>
-    operator_bounds_error_on_device(Map const&) {
-  Kokkos::abort("View bounds error");
-}
-
-template <class Map>
-KOKKOS_FUNCTION
-    std::enable_if_t<is_detected<printable_label_typedef_t, Map>::value>
-    operator_bounds_error_on_device(Map const& map) {
-  OperatorBoundsErrorOnDevice<Map>::run(map);
-}
-
 template <class MemorySpace, class ViewType, class MapType, class... Args>
 KOKKOS_INLINE_FUNCTION void view_verify_operator_bounds(
     Kokkos::Impl::ViewTracker<ViewType> const& tracker, const MapType& map,
     Args... args) {
-  if (!view_verify_operator_bounds<0>(map, args...)) {
-    KOKKOS_IF_ON_HOST(
-        (enum {LEN = 1024}; char buffer[LEN];
-         const std::string label =
-             tracker.m_tracker.template get_label<MemorySpace>();
-         int n = snprintf(buffer, LEN, "View bounds error of view %s (",
-                          label.c_str());
-         view_error_operator_bounds<0>(buffer + n, LEN - n, map, args...);
-         Kokkos::Impl::throw_runtime_exception(std::string(buffer));))
-
-    KOKKOS_IF_ON_DEVICE((
-        /* Check #1: is there a SharedAllocationRecord?
-           (we won't use it, but if its not there then there isn't
-            a corresponding SharedAllocationHeader containing a label).
-           This check should cover the case of Views that don't
-           have the Unmanaged trait but were initialized by pointer. */
-        if (tracker.m_tracker.has_record()) {
-          operator_bounds_error_on_device(map);
-        } else { Kokkos::abort("View bounds error"); }))
+  if (!check_bounds(map, std::make_index_sequence<sizeof...(Args)>(),
+                    args...)) {
+    char err[256] = "";
+    Kokkos::Impl::strcat(err, "Kokkos::View ERROR: out of bounds access");
+    Kokkos::Impl::strcat(err, " label=(\"");
+    KOKKOS_IF_ON_HOST([&] {
+      if (tracker.m_tracker.has_record()) {
+        Kokkos::Impl::strncat(
+            err, tracker.m_tracker.template get_label<void>().c_str(), 128);
+        return;
+      }
+      Kokkos::Impl::strcat(err, "**UNMANAGED**");
+    }();)
```

Comment:
We don't use immediately invoked lambdas elsewhere with `KOKKOS_IF_ON_HOST`. Did you introduce it because of the extra control flow (`return;`) used below?

Reply:
I introduced it to reduce the complexity of the control flow in the device side and I decided to use it as well for the host code where the improvement is less obvious but I judged that consistency with the device code was a good thing.

Reply:
here is what it looks like without the lambda trick
```diff
@@ -3580,13 +3580,12 @@ KOKKOS_INLINE_FUNCTION void view_verify_operator_bounds(
     char err[256] = "";
     strcat(err, "Kokkos::View ERROR: out of bounds access");
     strcat(err, " label=(\"");
-    KOKKOS_IF_ON_HOST([&] {
-      if (tracker.m_tracker.has_record()) {
-        strncat(err, tracker.m_tracker.template get_label<void>().c_str(), 128);
-        return;
-      }
-      strcat(err, "**UNMANAGED**");
-    }();)
+    KOKKOS_IF_ON_HOST(
+        if (tracker.m_tracker.has_record()) {
+          strncat(err, tracker.m_tracker.template get_label<void>().c_str(),
+                  128);
+          return;
+        } else { strcat(err, "**UNMANAGED**"); })
     KOKKOS_IF_ON_DEVICE([&] {
       // Check #1: is there a SharedAllocationRecord?  (we won't use it, but
       // if its not there then there isn't a corresponding

Reply:
Anyway I changed it since both Christian and you complained about it